R/cell_labeling.R

```diff
@@ -0,0 +1,223 @@
+#'   expressed genes as background to account for cell quality and
+#'   sequencing depth.
+#'
+#' @author Michael Stadler
+#'
+#' @param sce \code{SingleCellExperiment} object.
+#' @param genes \code{character} vector with the genes in the set. Must be a
+#'   subset of \code{rownames(sce)}.
+#' @param expr_values Integer scalar or string indicating which assay of
+#'   \code{sce} contains the expression values.
+#' @param R Integer scalar giving the number of random gene sets to sample
+#'   for normalization.
+#' @param BPPARAM An optional \code{\link[BiocParallel]{BiocParallelParam}}
+#'   instance determining the parallel back-end to be used during evaluation.
+#' 
+#' @return A \code{numeric} vector with normalized gene set scores for each
+#'   cell in \code{sce}.
+#' 
+#' @importFrom BiocParallel SerialParam bplapply
+#' @importFrom SummarizedExperiment assay assayNames
+#' @importFrom matrixStats rowMeans2 rowSds
+#' 
+#' @export
+normGenesetExpression <- function(sce,
+                                  genes,
+                                  expr_values = "logcounts",
+                                  R = 200,
+                                  BPPARAM = SerialParam()) {
+    # pre-flight checks
+    stopifnot(exprs = {
+        is(sce, "SingleCellExperiment")
+        is.character(genes)
+        length(genes) > 1L
+        all(genes %in% rownames(sce))
+        is.numeric(R)
+        length(R) == 1L
+        R > 0
+        inherits(BPPARAM, "BiocParallelParam")
+    })
+    if (is.numeric(expr_values) && length(expr_values) == 1L)
+        expr_values <- assayNames(sce)[expr_values]
+    stopifnot(exprs = {
+        is.character(expr_values)
+        length(expr_values) == 1L
+        expr_values %in% assayNames(sce)
+    })
+    
+    # calculate average expression of selected genes
+    expr <- assay(sce, expr_values)
+    i <- match(genes[genes %in% rownames(expr)], rownames(expr))
+    avgExpr <- rowMeans(expr)
+    val.obs <- colSums(expr[i, , drop = FALSE])
+    
+    # calculate expression in random gene sets
+    val.rand <- do.call(cbind, bplapply(seq_len(R), function(r) {
+        irand <- sampleControlElements(x = avgExpr, idxTarget = i, nbins = 100)
+        colSums(expr[irand, , drop = FALSE])
+    }, BPPARAM = BPPARAM))
+    
+    # normalize observed gene set expression
+    # z values: (obs - exp) / sqrt(Var[exp])
+    mus <- rowMeans2(val.rand)
+    sds <- rowSds(val.rand)
+    val.obs.norm <- (val.obs - mus) / sds
+    names(val.obs.norm) <- colnames(sce)
+    
+    return(val.obs.norm)
+}
+
+
+#' @title Assign labels to cells using known marker genes
+#' 
+#' @description Given marker gene sets for cell types, identify cells with
+#'   high expression of the marker genes (positive examples), then use these
+#'   cells to create a reference transcriptome profile for each cell type and
+#'   identify additional cells of each type using \code{SingleR}.
```

Comment:
I guess these marker genes should be as specific as possible, and including, let's say, generic T-cell markers in an all-T-cell data set might actually introduce some spurious (oppositely directed) signals in the calculation of normalized gene set scores?

Reply:
Good point - I add a sentence along those lines to the description.